janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/HasStepFolder.java

```diff
@@ -212,6 +223,41 @@ static boolean validFoldInHasContainer(final Step<?, ?>  tinkerpopStep, final bo
         return Boolean.TRUE.equals(toReturn);
     }
 
+    /**
+     * Check if a given step is an "exists" step, i.e. has(key) step
+     * Tinkerpop translates has(key) gremlin query into TraversalFilterStep(PropertiesStep)
+     * @param step
+     * @return true if given step is essentially an "exists" step
+     */
+    static boolean isExistsStep(final Step<?, ?> step) {
+        if (step instanceof TraversalFilterStep) {
+            List<Traversal.Admin> traversals = ((TraversalFilterStep) step).getLocalChildren();
+            assert (traversals.size() == 1);
+            List<Step> steps = traversals.get(0).getSteps();
+            if (!(steps.size() == 1 && steps.get(0) instanceof PropertiesStep)) return false;
+            return ((PropertiesStep) steps.get(0)).getPropertyKeys().length == 1;
+        }
+        return false;
+    }
+
+    /**
+     * Convert a TraversalFilterStep that is essentially "has(key)" into a normal HasStep
+     * @param traversal local traversal
+     * @param currentStep
+     */
+    static Step foldInHasFilter(final Traversal.Admin<?, ?> traversal, final TraversalFilterStep<?> currentStep) {
+        List<? extends Traversal.Admin<?, ?>> traversals = currentStep.getLocalChildren();
+        assert(traversals.size() == 1);
+        List<Step> steps = traversals.get(0).getSteps();
+        assert(steps.size() == 1 && steps.get(0) instanceof PropertiesStep);
+        String[] propertyKeys = ((PropertiesStep) steps.get(0)).getPropertyKeys();
+        assert(propertyKeys.length == 1);
+        HasStep hasStep = new HasStep(traversal, new HasContainer(propertyKeys[0], P.neq(null)));
```

Comment:
I'm not sure `.has("prop", P.neq(null))` and `.has("prop")` are equal. At least, in TinkerPop 3.5.0 `null` values are supported. 
That said, I didn't check this PR much. Maybe I just miss something

Reply:
This is the semantics JanusGraph has been following:
https://github.com/JanusGraph/janusgraph/blob/6785f762ba97424c1ab1cb5482b579c54c91d9f5/janusgraph-core/src/main/java/org/janusgraph/graphdb/query/graph/GraphCentricQueryBuilder.java#L132-L135

I read https://tinkerpop.apache.org/docs/3.5.0-SNAPSHOT/upgrade/#_null_semantics, but I don't think we are gonna support null values in the near future. There are some obstacles, for example, Cassandra does not support null values, unless we come up with some fixed custom value that can represent "null" values. If that happens, then we can translate `.has("prop")` into `.has("prop", P.neq(NULL_VALUE))` without many changes here.

Reply:
Great explanation @li-boxuan !
Agree, we don't need to support null values as for now because it is an optional feature in TinkerPop